### PR TITLE
Adding functionality for switching between counsel-file-jump and counsel-find-file

### PR DESCRIPTION
**Pull request by ngharrison**
Please take a look at these changes. First, they allow the user to to open counsel-file-jump from whatever directory they are currently viewing in counsel-find-file. Second, they add functionality to jump back to counsel-find-file from counsel-file-jump. This supplies a simple, interactive way to browse through and search within directories.

I followed the pattern of existing code for my edits and have run
```
make compile
make deps
make test
make checkdoc
```
without errors.

### CODE REVIEW

File: counsel.el [hunk 1]
```diff
@@ -1812,11 +1812,25 @@ currently checked out."
     (define-key map [remap undo] 'counsel-find-file-undo)
     map))
 
+;;* File
+;;** `counsel-file-jump'
+(defvar counsel-file-jump-map
+  (let ((map (make-sparse-keymap)))
+    (define-key map (kbd "`") #'counsel-file-find-from-jump)
+    map))
```
Comment (basil-conto) [inline counsel.el:1820]: Nit: Please always add docstrings. (I know a lot of existing code lacks them, but let's not proliferate that.)

File: counsel.el [hunk 2]
```diff
@@ -1812,11 +1812,25 @@ currently checked out."
...
+
 (defun counsel-file-jump-from-find ()
   "Switch to `counsel-file-jump' from `counsel-find-file'."
   (interactive)
   (ivy-quit-and-run
-    (counsel-file-jump ivy-text)))
+    (counsel-file-jump ivy-text (ivy-state-directory ivy-last))))
+
+(defun counsel-file-find-from-jump ()
```
Comment (basil-conto) [inline counsel.el:1828]: Shouldn't this be called `counsel-find-file-from-jump`?

File: counsel.el [hunk 3]
```diff
@@ -1812,11 +1812,25 @@ currently checked out."
...
+  "Switch to `counsel-find-file' from `counsel-file-jump'."
+  (interactive)
+  (ivy-quit-and-run
+    (counsel-find-file ivy-text (ivy-state-directory ivy-last)))
+  )
```
Comment (basil-conto) [inline counsel.el:1833]: Nit: Please don't leave hanging closing parens, it makes Lispers nervous. :)

File: counsel.el [hunk 4]
```diff
@@ -2011,11 +2025,11 @@ The preselect behavior can be customized via user options
...
 
-(defun counsel--find-file-1 (prompt initial-input action caller)
+(defun counsel--find-file-1 (prompt initial-input action caller &optional initial-directory)
   (let ((default-directory
-         (if (eq major-mode 'dired-mode)
-             (dired-current-directory)
-           default-directory)))
+          (if (eq major-mode 'dired-mode)
+              (dired-current-directory)
+            (or initial-directory default-directory))))
```
Comment (basil-conto) [inline counsel.el:2032]: Why the change in indentation?

File: counsel.el [hunk 5]
```diff
@@ -2027,14 +2041,15 @@ The preselect behavior can be customized via user options
...
+(defun counsel-find-file (&optional initial-input initial-directory)
   "Forward to `find-file'.
 When INITIAL-INPUT is non-nil, use it in the minibuffer during completion."
   (interactive)
   (counsel--find-file-1
    "Find file: " initial-input
    #'counsel-find-file-action
-   'counsel-find-file))
+   'counsel-find-file
+   initial-directory))
```
Comment (basil-conto) [inline counsel.el:2052]: Instead of passing this as a new argument to `counsel--find-file-1`, why not bind `default-directory` around the call to it?

**Review by basil-conto — COMMENTED** [review]
Thanks, just some minor comments from me.

File: counsel.el (hunk 2 above)
Comment (ngharrison) [reply counsel.el]: I now understand the naming convention. I have fixed that.

File: counsel.el (hunk 4 above)
Comment (ngharrison) [reply counsel.el]: I didn't catch that. Thanks.

File: counsel.el (hunk 5 above)
Comment (ngharrison) [reply counsel.el:2052]: That sounds like a fine idea to me, though I don't know how to go about doing that. Any direction on this?

Comment (basil-conto) [reply counsel.el:2052]: Here's what I mean:

```diff
diff --git a/counsel.el b/counsel.el
index 0e6a26f2..34292c58 100644
--- a/counsel.el
+++ b/counsel.el
@@ -2031,10 +2031,10 @@ counsel-find-file
   "Forward to `find-file'.
 When INITIAL-INPUT is non-nil, use it in the minibuffer during completion."
   (interactive)
-  (counsel--find-file-1
-   "Find file: " initial-input
-   #'counsel-find-file-action
-   'counsel-find-file))
+  (let ((default-directory initial-directory))
+    (counsel--find-file-1 "Find file: " initial-input
+                          #'counsel-find-file-action
+                          'counsel-find-file)))
 
 (ivy-configure 'counsel-find-file
   :parent 'read-file-name-internal
```

Comment (ngharrison) [reply counsel.el]: I see better what you mean, looks good to me.

File: counsel.el [hunk 6]
```diff
@@ -1812,11 +1812,25 @@ currently checked out."
     (define-key map [remap undo] 'counsel-find-file-undo)
     map))
 
+;;* File
+;;** `counsel-file-jump'
```
Comment (basil-conto) [inline counsel.el:1816]: These headings already exist in this file, and should be reused/fixed as required.

**Review by basil-conto — COMMENTED** [review]
Thanks, LGTM apart from a minor comment.

I'm not sure if I can merge this many changes without a copyright assignment to the FSF, though. Do you already have one on file? If not, would you be willing to submit one? If you're not familiar with this, see [`(info "(emacs) Copyright Assignment")`](https://www.gnu.org/software/emacs/manual/html_node/emacs/Copyright-Assignment.html). TIA.

File: counsel.el (hunk 6 above)
Comment (ngharrison) [reply counsel.el:1816]: I'd love to fix it, but I'm not sure what the file/comment conventions are. I may just as well mess it up again as correct it. Could you explain them? Or is this something you can do after the fact?

**Comment by ngharrison** [comment]
This is the first time I've ever contributed to Emacs/Elisp, so I have no such file. Is this necessary? If so, then I'll have to return to this some other time.

File: counsel.el (hunk 6 above)
Comment (basil-conto) [reply counsel.el:1816]: I would normally push these changes onto your branch before merging, but as a Collaborator I don't think I have the permissions to do so.

The comments in `counsel.el` are structured like an Org-Mode tree of outlines:
```
* Top-level heading
** Sub-heading
*** Sub-sub-heading
** Another sub-heading
* Another top-level heading
```

`counsel.el` has the following top-level groups:
```
* Utility
* Async Utility
* Completion at point
* Elisp symbols
* Git
* File
* Grep
* Org
* Misc. OS
* `counsel-mode'
```

Under `* File`, you will also find `** counsel-find-file` and `** counsel-file-jump`. My suggestion is that you place your new code under the appropriate heading, and move existing headings to the right place if necessary.

**Comment by basil-conto** [comment]
> This is the first time I've ever contributed to Emacs/Elisp, so I have no such file. Is this necessary? If so, then I'll have to return to this some other time.

The CA is generally needed for changes of more than ~15 lines. I'm not sure if this PR qualifies as under the limit. If you're willing to go through with it, the process is very simple: just fill in the form at https://github.com/coreutils/gnulib/blob/master/doc/Copyright/request-assign.future and email it to assign@gnu.org, who will guide you from there. TIA!

File: counsel.el [hunk 7]
```diff
@@ -2864,8 +2864,21 @@ INITIAL-DIRECTORY, if non-nil, is used as the root directory for search."
               :preselect (counsel--preselect-file)
               :require-match 'confirm-after-completion
               :history 'file-name-history
+              :keymap counsel-file-jump-map
```
Comment (basil-conto) [inline counsel.el:2867]: The variable is not yet defined here, giving rise to the following byte-compilation warning:
```
In counsel-file-jump:
counsel.el:2867:23: Warning: reference to free variable
    ‘counsel-file-jump-map’
```
The solution is to move the variable's definition before to a point before its first use.

Comment (ngharrison) [reply counsel.el:2879]: Ok, I hope that's fixed now.

Comment (basil-conto) [reply counsel.el:2879]: Yes, thanks.

**Review by basil-conto — APPROVED** [review]
LGTM, so just waiting for the CA now.

**Comment by ngharrison** [comment]
I believe all the appropriate adjustments are now made. I have also submitted a copyright assignment to GNU.

**Comment by basil-conto** [comment]
> I have also submitted a copyright assignment to GNU.

Your CA doesn't seem to be on file. Have you completed the assignment process, or did you only just submit the request? In the former case, maybe it takes a little while for the list to be updated.

**Comment by ngharrison** [comment]
It took a couple weeks for fsf.org to get back to me. They said the process is complete, so I hope it shows up now.

**Comment by basil-conto** [comment]
Thanks! Just rebased off `master`, squashed the commits with an expanded commit message, and merged.
